src/labels_transfer/pynndescent_knn/script.py

```diff
@@ -0,0 +1,160 @@
+import mudata as mu
+import numba
+import numpy as np
+from scipy.sparse import issparse
+import sys
+from pynndescent import PyNNDescentTransformer
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.pipeline import make_pipeline
+
+## VIASH START
+par = {
+    "input": "resources_test/pbmc_1k_protein_v3/pbmc_1k_protein_v3_mms.h5mu",
+    "modality": "rna",
+    "input_obsm_features": None,
+    "reference": "resources_test/annotation_test_data/TS_Blood_filtered.h5mu",
+    "reference_obsm_features": None,
+    "reference_obs_targets": ["cell_type"],
+    "output": "foo.h5mu",
+    "output_obs_predictions": None,
+    "output_obs_probability": None,
+    "output_uns_parameters": "labels_transfer",
+    "output_compression": None,
+    "weights": "uniform",
+    "n_neighbors": 15
+}
+meta ={
+    "resources_dir": "src/labels_transfer/utils"
+}
+## VIASH END
+
+sys.path.append(meta["resources_dir"])
+from helper import check_arguments, get_reference_features, get_query_features, check_sparsity
+
+def setup_logger():
+    import logging
+    from sys import stdout
+
+    logger = logging.getLogger()
+    logger.setLevel(logging.INFO)
+    console_handler = logging.StreamHandler(stdout)
+    logFormatter = logging.Formatter("%(asctime)s %(levelname)-8s %(message)s")
+    console_handler.setFormatter(logFormatter)
+    logger.addHandler(console_handler)
+
+    return logger
+# END TEMPORARY WORKAROUND setup_logger
+
+@numba.njit
+def weighted_prediction(weights, ref_cats):
+    """Get highest weight category."""
+    N = len(weights)
+    predictions = np.zeros((N,), dtype=ref_cats.dtype)
+    probabilities = np.zeros((N,))
+    for i in range(N):
+        obs_weights = weights[i]
+        obs_cats = ref_cats[i]
+        best_prob = 0
+
+        neighbors_categories = np.unique(obs_cats)
+        
+        # If all neighbors have the same class, just predict it. This saves quite some time
+        if len(neighbors_categories) == 1:
+            predictions[i] = neighbors_categories[0]
+            probabilities[i] = 1
+        else:
+            for c in neighbors_categories:
+                cand_prob = np.sum(obs_weights[obs_cats == c])
+                if cand_prob > best_prob:
+                    best_prob = cand_prob
+                    predictions[i] = c
+                    probabilities[i] = best_prob
+
+    return predictions, probabilities
+
+def distances_to_affinities(distances):
+    stds = np.std(distances, axis=1)
+    stds = (2.0 / stds) ** 2
+    stds = stds.reshape(-1, 1)
+    distances_tilda = np.exp(-np.true_divide(distances, stds))
+
+    return distances_tilda / np.sum(distances_tilda, axis=1, keepdims=True)
+
+logger = setup_logger()
+
+# Reading in data
+logger.info(f"Reading in query dataset {par['input']} and reference datasets {par['reference']}")
+q_mdata = mu.read_h5mu(par["input"])
+q_adata = q_mdata.mod[par["modality"]]
```

Comment:
Question: do you add the output directly on the input object? Otherwise I would prefer to provide a minimal adata as input and copy over the output slots to the input adata

Reply:
Yes, we save the output to the same input object by adding `.obs` columns with labels